### PR TITLE
feat: enable support from/to pojo/protobuf for extended expressions

### PR DESCRIPTION
**Pull request by davisusanibar**
To closes https://github.com/substrait-io/substrait-java/issues/127

### CODE REVIEW

File: core/src/main/java/io/substrait/extended/expression/ExtendedExpressionProtoConverter.java
Comment (danepitkin) [inline core/src/main/java/io/substrait/extended/expression/ExtendedExpressionProtoConverter.java:18]:
```suggestion
    ExtendedExpression.Builder builder = ExtendedExpression.newBuilder();
```
nit: this is just for readability

File: core/src/main/java/io/substrait/extended/expression/ProtoExtendedExpressionConverter.java
Comment (danepitkin) [inline core/src/main/java/io/substrait/extended/expression/ProtoExtendedExpressionConverter.java:5]: Can we explicitly list imports instead of using `*`?

File: core/src/main/java/io/substrait/extended/expression/ProtoExtendedExpressionConverter.java
Comment (danepitkin) [inline core/src/main/java/io/substrait/extended/expression/ProtoExtendedExpressionConverter.java:70]:
```suggestion
  private io.substrait.type.NamedStruct convertNamedStructProtoToPojo(NamedStruct namedStruct) {
```

File: core/src/main/java/io/substrait/extension/ImmutableExtensionLookup.java
Comment (danepitkin) [inline core/src/main/java/io/substrait/extension/ImmutableExtensionLookup.java:34]: Nice refactor! we could also overload the API and add `from(Plan p)` and `from(ExtendedExpression e)` that call `from(simpleExtensionURIs, simpleExtensionDeclarations)`.

File: core/src/main/java/io/substrait/extended/expression/ProtoExtendedExpressionConverter.java
Comment (danepitkin) [reply core/src/main/java/io/substrait/extended/expression/ProtoExtendedExpressionConverter.java:70]: Would it make sense to implement this function as `fromProto` in this file? https://github.com/substrait-io/substrait-java/blob/main/core/src/main/java/io/substrait/type/NamedStruct.java

**Review by danepitkin — COMMENTED** [review]
This looks awesome! I took a first pass at the feature, but haven't reviewed the tests yet.

File: core/src/test/java/io/substrait/extended/expression/ProtoExtendedExpressionConverterTest.java
Comment (vbarua) [inline core/src/test/java/io/substrait/extended/expression/ProtoExtendedExpressionConverterTest.java:36]: Could you create the expression here explicitly rather than pulling it out from the extensions like this. I'm not a fan of this pattern because I think:
1. It makes the test harder to read. There's a lot of extra code to pick out one declaration rather than creating one.
2. It makes the test depend on the _ordering_ of declarations in the default collection. There are 13 add variants and the first one is the decimal one. If the ordering changes though (and it's not guaranteed to be stable) the test will start to fail randomly.

There's is a 
```java
protected SubstraitBuilder b
```
in the `TestBase` class with a `scalarFunction` method that can be used for this.

File: core/src/main/java/io/substrait/extended/expression/ExtendedExpression.java
Comment (vbarua) [reply core/src/main/java/io/substrait/extended/expression/ExtendedExpression.java:1]: I like this idea as well. Having an `.../extended` directory makes me think there could be more than one item in it. `extendedexpression` maps to a singular `ExtensionExpression` better, at least to me.

File: core/src/main/java/io/substrait/extension/ImmutableExtensionLookup.java
Comment (vbarua) [reply core/src/main/java/io/substrait/extension/ImmutableExtensionLookup.java:34]: +1 to keeping the existing `from(Plan p)` and having it call your newly refactor method.

File: core/src/main/java/io/substrait/extended/expression/ExtendedExpression.java
Comment (davisusanibar) [reply core/src/main/java/io/substrait/extended/expression/ExtendedExpression.java]: Changed

File: core/src/main/java/io/substrait/extended/expression/ExtendedExpression.java
Comment (davisusanibar) [reply core/src/main/java/io/substrait/extended/expression/ExtendedExpression.java]: I have added AggregateFunction and marked it as an UnsupportedOperationException for now.

File: core/src/main/java/io/substrait/extended/expression/ProtoExtendedExpressionConverter.java
Comment (davisusanibar) [reply core/src/main/java/io/substrait/extended/expression/ProtoExtendedExpressionConverter.java]: This is the result of applying `./gradlew spotlessApply` let me check if there are changes needed

File: core/src/main/java/io/substrait/extended/expression/ProtoExtendedExpressionConverter.java
Comment (davisusanibar) [reply core/src/main/java/io/substrait/extended/expression/ProtoExtendedExpressionConverter.java]: Changed

File: core/src/main/java/io/substrait/extension/ImmutableExtensionLookup.java
Comment (davisusanibar) [reply core/src/main/java/io/substrait/extension/ImmutableExtensionLookup.java]: Added

File: core/src/test/java/io/substrait/extended/expression/ProtoExtendedExpressionConverterTest.java
Comment (davisusanibar) [reply core/src/test/java/io/substrait/extended/expression/ProtoExtendedExpressionConverterTest.java]: Added

File: core/src/test/java/io/substrait/extendedexpression/ProtoExtendedExpressionConverterTest.java
Comment (danepitkin) [inline core/src/test/java/io/substrait/extendedexpression/ProtoExtendedExpressionConverterTest.java:22]: It might be better to have a single `roundtripExtendedExpression` test so that you can test both to/from proto and compare the full objects afterwards (instead of `toProtoTest` and `fromProtoTest`).

**Review by danepitkin — COMMENTED** [review]
Thanks David! I left a few more comments

File: core/src/main/java/io/substrait/type/NamedStruct.java
Comment (davisusanibar) [reply core/src/main/java/io/substrait/type/NamedStruct.java]: Changed

File: core/src/main/java/io/substrait/extendedexpression/ExtendedExpressionProtoConverter.java
Comment (davisusanibar) [reply core/src/main/java/io/substrait/extendedexpression/ExtendedExpressionProtoConverter.java]: Changed

File: core/src/main/java/io/substrait/extendedexpression/ExtendedExpressionProtoConverter.java
Comment (davisusanibar) [reply core/src/main/java/io/substrait/extendedexpression/ExtendedExpressionProtoConverter.java]: Just added.

File: core/src/main/java/io/substrait/extendedexpression/ProtoExtendedExpressionConverter.java
Comment (davisusanibar) [reply core/src/main/java/io/substrait/extendedexpression/ProtoExtendedExpressionConverter.java]: Changed.

Comment (davisusanibar) [reply core/src/main/java/io/substrait/extendedexpression/ProtoExtendedExpressionConverter.java]: Solved by applying https://github.com/substrait-io/substrait-java/pull/191#discussion_r1408378391

File: core/src/test/java/io/substrait/extendedexpression/ExtendedExpressionProtoConverterTest.java
Comment (davisusanibar) [reply core/src/test/java/io/substrait/extendedexpression/ExtendedExpressionProtoConverterTest.java]: These variables are private in SubstraitBuilder, so just maintain them here for testing purposes.

File: core/src/test/java/io/substrait/extendedexpression/ProtoExtendedExpressionConverterTest.java
Comment (davisusanibar) [reply core/src/test/java/io/substrait/extendedexpression/ProtoExtendedExpressionConverterTest.java]: Just maintain only roundtrip class for Expression, Aggregation and Expression + Aggregation

File: core/src/main/java/io/substrait/extendedexpression/ProtoExtendedExpressionConverter.java
Comment (danepitkin) [inline core/src/main/java/io/substrait/extendedexpression/ProtoExtendedExpressionConverter.java:52]: optional: is there another provided option that's better than comparing with `getNumber()`? If not, then I think this is okay.

File: core/src/test/java/io/substrait/extendedexpression/ExtendedExpressionRoundTripTest.java
Comment (danepitkin) [inline core/src/test/java/io/substrait/extendedexpression/ExtendedExpressionRoundTripTest.java:19]: nit: could we put this here instead? https://github.com/substrait-io/substrait-java/blob/e73c6cac7a5b62d026ccd13eca97bfae02395aab/core/src/main/java/io/substrait/dsl/SubstraitBuilder.java#L48

File: core/src/test/java/io/substrait/extendedexpression/ExtendedExpressionRoundTripTest.java
Comment (danepitkin) [inline core/src/test/java/io/substrait/extendedexpression/ExtendedExpressionRoundTripTest.java:1]: optional nit: the tests might be more readable with helper functions for `createTestExpressionReference()`, `createTestAggregateReference()` and `verifyRoundTrip()`. It could help reduce code duplication, too.

**Review by danepitkin — COMMENTED** [review]
Nice work, @davisusanibar ! Left a few minor comments, but overall LGTM.

File: core/src/main/java/io/substrait/extendedexpression/ExtendedExpressionProtoConverter.java
Comment (vbarua) [inline core/src/main/java/io/substrait/extendedexpression/ExtendedExpressionProtoConverter.java:11]: minor: As is these docs links are the same class. If we fully qualify them it's clearer that they are distinct ExtendedExpression classes.
java
```
/** Converts from {@link io.substrait.extendedexpression.ExtendedExpression} to {@link io.substrait.proto.ExtendedExpression} */
```

File: core/src/main/java/io/substrait/extendedexpression/ExtendedExpressionProtoConverter.java
Comment (vbarua) [inline core/src/main/java/io/substrait/extendedexpression/ExtendedExpressionProtoConverter.java:27]: I made a small change here (https://github.com/substrait-io/substrait-java/pull/206/commits/92d2cc57e46e4b4972e5251477abf6d9baf4300b) to bind the variables your instanceof checking to make the PR easier to read. Using this, this way you can avoid the need to recast them.

File: core/src/main/java/io/substrait/extendedexpression/ExtendedExpression.java
Comment (vbarua) [inline core/src/main/java/io/substrait/extendedexpression/ExtendedExpression.java:25]: I think that having the ExpressionReference contain the type of the reference within itself is a bit clunky as a design. Looking at the `ExtendedExpressionProtoConverter`, we have to do instanceof checks to figure out the type and then make decision based on the type.

I think a cleaner way to approach this is with inheritance with something like:
```java
  public interface ExpressionReferenceBase {
    public List<String> getOutputNames();

    public io.substrait.proto.ExpressionReference toProto(...);
  }

  @Value.Immutable
  public abstract static class ExpressionReference implements ExpressionReferenceBase {
    public abstract Expression getExpression();

    @Override
    public io.substrait.proto.ExpressionReference toProto(...) {
      return ...
    }
  }

  @Value.Immutable
  public abstract static class AggregateFunctionReference implements ExpressionReferenceBase {
    public abstract Aggregate.Measure getMeasure();

    @Override
    public io.substrait.proto.ExpressionReference toProto(ExpressionProtoConverter converter) {
      return null;
    }
  }
```
effectively, each class knows how to convert itself.

@davisusanibar what do you think?

**Comment by danepitkin** [comment]
Thanks, Victor, for pointing out there is still some work to do to implement AggregateFunctions. It seems like it may be best to push AggregateFunction support to a follow up issue to keep this PR small.

File: core/src/main/java/io/substrait/extendedexpression/ExtendedExpressionProtoConverter.java
Comment (vbarua) [reply core/src/main/java/io/substrait/extendedexpression/ExtendedExpressionProtoConverter.java]: I think this actually may be a fair bit of work to do properly. In the interest of keeping this PR small and moving this work along, I think we could include AggregateFunction support as a future change.

What do you think @danepitkin

File: core/src/test/java/io/substrait/extendedexpression/ExtendedExpressionRoundTripTest.java
Comment (danepitkin) [reply core/src/test/java/io/substrait/extendedexpression/ExtendedExpressionRoundTripTest.java:19]: Nevermind, these are private so lets ignore this comment for now.

File: core/src/main/java/io/substrait/extendedexpression/ExtendedExpression.java
Comment (davisusanibar) [reply core/src/main/java/io/substrait/extendedexpression/ExtendedExpression.java]: I'm sorry, that was my mistake. Just update the code to use POJOs instead of PROTOs.

**Comment by davisusanibar** [comment]
> Thanks, Victor, for pointing out there is still some work to do to implement AggregateFunctions. It seems like it may be best to push AggregateFunction support to a follow up issue to keep this PR small.

Hi Dane, just doing my last attempt to try to cover AggregateFunction ... I could suggest that it is needed to have a uniform API that won't change that dramatically later.

File: core/src/main/java/io/substrait/extendedexpression/ExtendedExpressionProtoConverter.java
Comment (vbarua) [inline core/src/main/java/io/substrait/extendedexpression/ExtendedExpressionProtoConverter.java:31]: Why are we restricting this to scalar functions invocations? Are the other expression types not valid here?

File: core/src/main/java/io/substrait/extendedexpression/ProtoExtendedExpressionConverter.java
Comment (vbarua) [reply core/src/main/java/io/substrait/extendedexpression/ProtoExtendedExpressionConverter.java:53]: IMO the nicer way of doing this is with a switch statement like
```java
      switch (expressionReference.getExprTypeCase()) {
        case EXPRESSION -> {
         ...
        }
        case MEASURE -> {
        ...
        }
        default -> throw new UnsupportedOperationException(
            "Only Expression or Aggregate Function type are supported in conversion from proto Extended Expressions");
      }
```

which avoids the magic number problem and is easier to read than if-else-if (at least for me).

File: core/src/main/java/io/substrait/extendedexpression/ExtendedExpression.java
Comment (vbarua) [inline core/src/main/java/io/substrait/extendedexpression/ExtendedExpression.java:21]: This needs to be the POJO AdvancedExtension not the proto one.

Comment (vbarua) [reply core/src/main/java/io/substrait/extendedexpression/ExtendedExpression.java:21]: Wait, the Plan POJO also use the protobuf AdvancedExtension.

https://github.com/substrait-io/substrait-java/blob/e73c6cac7a5b62d026ccd13eca97bfae02395aab/core/src/main/java/io/substrait/plan/Plan.java#L3C2-L3C45

I'll file an issue to fix this in the future.

Comment (vbarua) [reply core/src/main/java/io/substrait/extendedexpression/ExtendedExpression.java:21]: Filed: https://github.com/substrait-io/substrait-java/issues/210

File: core/src/main/java/io/substrait/extendedexpression/ExtendedExpressionProtoConverter.java
Comment (davisusanibar) [reply core/src/main/java/io/substrait/extendedexpression/ExtendedExpressionProtoConverter.java:31]: Ah, I see, I'm only considering Filter and Projection cases. Let me review changes needed to be more generic.

Comment (davisusanibar) [reply core/src/main/java/io/substrait/extendedexpression/ExtendedExpressionProtoConverter.java]: Changed

File: core/src/main/java/io/substrait/extendedexpression/ProtoExtendedExpressionConverter.java
Comment (danepitkin) [reply core/src/main/java/io/substrait/extendedexpression/ProtoExtendedExpressionConverter.java:53]: +1

**Comment by danepitkin** [comment]
I'll let @vbarua comment on if there is more work needed for AggregateFunctions, as I'm not yet that familiar with them. If there is more work required, we should do it in a follow up PR to keep this one small. The Expression support LGTM and I wouldn't want to hold up merging it, especially if the SQL -> ExtendedExpression PR would benefit from it!

File: core/src/main/java/io/substrait/relation/AggregateFunctionProtoConverter.java
Comment (vbarua) [inline core/src/main/java/io/substrait/relation/AggregateFunctionProtoConverter.java:42]: This duplicates code `RelProtoConverter#toProto(Aggregate.Measure measure)`.

It's fine for now, but we should consider unifying it in the future.

File: core/src/test/java/io/substrait/extendedexpression/ExtendedExpressionRoundTripTest.java
Comment (vbarua) [inline core/src/test/java/io/substrait/extendedexpression/ExtendedExpressionRoundTripTest.java:32]: TIL about `@MethodSource`. That's pretty nifty.

**Review by vbarua — COMMENTED** [review]
Thanks for adding this. I left some comments about things that we can improve in the future, but overall looks great.

I would like two thing before approving and merging:
1. Could you update the `ProtoExtendedExpressionConverter` to not use comparisons to `getNumber` and instead compare the protobuf enums.
2. Could you answer my question about the `@Desugar` annotation.

File: core/src/main/java/io/substrait/relation/AggregateFunctionProtoConverter.java
Comment (davisusanibar) [reply core/src/main/java/io/substrait/relation/AggregateFunctionProtoConverter.java:41]: I will create a ticket to address this issue.

File: core/src/main/java/io/substrait/relation/ProtoAggregateFunctionConverter.java
Comment (davisusanibar) [reply core/src/main/java/io/substrait/relation/ProtoAggregateFunctionConverter.java:59]: I will create a ticket to address this issue.

File: core/src/main/java/io/substrait/extension/ExtensionCollector.java
Comment (davisusanibar) [reply core/src/main/java/io/substrait/extension/ExtensionCollector.java:115]: It might be possible to pass this Jabel validation https://github.com/bsideup/jabel/blob/master/jabel-javac-plugin/src/main/java/com/github/bsideup/jabel/RecordsRetrofittingTaskListener.java#L484

File: core/src/main/java/io/substrait/extendedexpression/ExtendedExpressionProtoConverter.java
Comment (davisusanibar) [reply core/src/main/java/io/substrait/extendedexpression/ExtendedExpressionProtoConverter.java:28]: I will create a ticket to address this issue.

**Review by vbarua — APPROVED** [review]
Changes look good. Thanks for working with us on this @davisusanibar!

**Comment by danepitkin** [comment]
Thanks, @davisusanibar ! Nice work!!

**Pull request merged**
